storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+
+    uint32_t new_page_no= new_block->page.id().page_no();
+    mtr.write<4,mtr_t::FORCED>(*new_block,
+                               FIL_PAGE_OFFSET + new_block->page.frame,
+                               new_page_no);
+
+    uint32_t prev_page_no= btr_page_get_prev(page);
+
+    /* change the previous page */
+    if (prev_page_no != FIL_NULL)
+    {
+      buf_block_t *prev_block=
+        buf_page_get_gen(page_id_t{0, prev_page_no}, 0, RW_X_LATCH,
+                         nullptr, BUF_GET_POSSIBLY_FREED, &mtr,
+                         &err);
+      if (!prev_block)
+        goto err_exit;
+      mtr.write<4>(*prev_block,
+                   prev_block->page.frame + FIL_PAGE_NEXT,
+                   new_page_no);
+    }
+
+    /* Change the next page */
+    if (next_page_no != FIL_NULL)
+    {
+      buf_block_t *next_block=
+        buf_page_get_gen(page_id_t{0, next_page_no}, 0, RW_X_LATCH,
+                         nullptr, BUF_GET_POSSIBLY_FREED, &mtr,
+                         &err);
+      if (!next_block)
+        goto err_exit;
+      mtr.write<4>(*next_block,
+                   next_block->page.frame + FIL_PAGE_PREV,
+                   new_page_no);
+    }
+
+    DBUG_EXECUTE_IF("fail_level_defragment",
+                    err= DB_CORRUPTION; goto err_exit;);
+    /* Change the parent page */
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it != m_parent_pages.end())
+    {
+      buf_block_t *parent_block=
+        buf_page_get_gen(page_id_t{0, uint32_t(parent_it->second)},
+                         0, RW_X_LATCH, nullptr,
+                         BUF_GET_POSSIBLY_FREED, &mtr, &err);
+      if (!parent_block)
+        goto err_exit;
+      mtr.write<4>(*parent_block,
+                    parent_block->page.frame
+                      + uint32_t(parent_it->second >> 32),
+                    new_page_no);
+    }
+
+    /* Free the page from the segment */
+    err= fseg_free_page(seg_header, space, cur_page_no, &mtr);
+    if (err)
+      goto err_exit;
+
+    if (state == XDES_FREE_FRAG || state == XDES_FULL_FRAG)
+    {
+      /* Put the new page in the fragment page array of the segment */
+      ulint n= fseg_find_free_frag_page_slot(inode);
+      if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+      {
+        err= DB_ERROR;
+        goto err_exit;
+      }
+      fseg_set_nth_frag_page_no(inode, iblock, n, new_page_no, &mtr);
+    }
+
+    block= new_block;
+    goto next_page;
+  }
+
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag)
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  mtr.start();
+  mtr.x_lock_space(fil_system.sys_space);
+  buf_block_t *root= buf_page_get_gen(page_id_t{0, m_index->page},
+                                      0, RW_SX_LATCH, nullptr,
+                                      BUF_GET_POSSIBLY_FREED, &mtr,
+                                      &err);
+  if (err)
+  {
+    mtr.commit();
+    return err;
+  }
```

Comment:
It would be more efficient to check `if (!root)` because that value must be readily already in the return value register in most if not all ABI and ISA.